File: src/python/tests/test_phonemize.py
"""
Unified phonemization tests - combines all phonemization testing
"""

import pytest

# Try to import implementation, skip if not available
pytest.importorskip("piper_train.phonemize")

from piper_train.phonemize.token_mapper import TOKEN2CHAR, CHAR2TOKEN, map_sequence

# Japanese imports are optional
try:
    import pyopenjtalk
    from piper_train.phonemize.japanese import phonemize_japanese

    HAS_JAPANESE = True
except ImportError:
    HAS_JAPANESE = False


class TestPhonemization:
    """All phonemization tests in one place"""

    @pytest.mark.unit
    def test_token_mapper_pua_mappings(self):
        """Test PUA character mappings are correct"""
        # Critical mappings for Japanese
        assert TOKEN2CHAR["ch"] == "\ue00e"
        assert TOKEN2CHAR["ts"] == "\ue00f"
        assert TOKEN2CHAR["ky"] == "\ue006"

        # Verify bidirectional mapping
        for token, char in TOKEN2CHAR.items():
            assert CHAR2TOKEN[char] == token

    @pytest.mark.unit
    def test_map_sequence(self):
        """Test phoneme sequence mapping"""
        input_seq = ["k", "o", "n", "n", "i", "ch", "i", "w", "a"]
        mapped = map_sequence(input_seq)

        # "ch" should be mapped to PUA
        assert mapped[5] == "\ue00e"
        # Others should remain unchanged
        assert mapped[0] == "k"
        assert mapped[6] == "i"

    @pytest.mark.unit
    @pytest.mark.japanese
    @pytest.mark.requires_openjtalk
    def test_japanese_basic(self):
        """Test basic Japanese phonemization"""
        if not HAS_JAPANESE:
            pytest.skip("Japanese phonemizer not available")

        # Test hiragana
        phonemes = phonemize_japanese("あ")
        assert "^" in phonemes  # Start marker
        assert "a" in phonemes  # Phoneme
        assert "$" in phonemes  # End marker

        # Test with multi-char phonemes
        phonemes = phonemize_japanese("ちゃ")
        assert len(phonemes) > 2

    @pytest.mark.unit
    def test_empty_input(self):
        """Test empty input handling"""
        # Empty list
        result = map_sequence([])
        assert result == []

File: src/cpp/tests/test_phonemize.cpp
/**
 * Unit tests for phonemization functionality
 */

#include <gtest/gtest.h>
#include <string>
#include <vector>
#include <map>

// Mock the phonemize functions for testing
namespace piper {

// Mock PUA mappings for testing
std::map<std::string, char32_t> testMultiCharToPUA = {
    {"ch", 0xE00E},
    {"ts", 0xE00F},
    {"ky", 0xE006},
    {"sh", 0xE010}
};

// Mock phoneme mapping function
std::vector<std::string> mapPhonemes(const std::vector<std::string>& phonemes) {
    std::vector<std::string> mapped;

    for (const auto& phoneme : phonemes) {
        auto it = testMultiCharToPUA.find(phoneme);
        if (it != testMultiCharToPUA.end()) {
            // Convert to UTF-8 string
            char32_t codepoint = it->second;
            if (codepoint <= 0x7F) {
                mapped.push_back(std::string(1, static_cast<char>(codepoint)));
            } else if (codepoint <= 0x7FF) {
                mapped.push_back(std::string{
                    static_cast<char>(0xC0 | (codepoint >> 6)),
                    static_cast<char>(0x80 | (codepoint & 0x3F))
                });
            } else if (codepoint <= 0xFFFF) {
                mapped.push_back(std::string{
                    static_cast<char>(0xE0 | (codepoint >> 12)),
                    static_cast<char>(0x80 | ((codepoint >> 6) & 0x3F)),
                    static_cast<char>(0x80 | (codepoint & 0x3F))
                });
            }
        } else {
            mapped.push_back(phoneme);
        }
    }

    return mapped;
}

} // namespace piper

// Test basic phoneme mapping
TEST(PhonemizeTest, BasicPhonemeMapping) {
    std::vector<std::string> input = {"a", "ch", "i"};
    auto result = piper::mapPhonemes(input);

    EXPECT_EQ(result.size(), 3);
    EXPECT_EQ(result[0], "a");
    EXPECT_NE(result[1], "ch"); // Should be mapped to PUA
    EXPECT_EQ(result[2], "i");
}

// Test all multi-char phoneme mappings
TEST(PhonemizeTest, AllMultiCharMappings) {
    std::vector<std::string> multiCharPhonemes = {"ch", "ts", "ky", "sh"};
    auto result = piper::mapPhonemes(multiCharPhonemes);

    EXPECT_EQ(result.size(), multiCharPhonemes.size());

    // All should be mapped (not equal to original)
    for (size_t i = 0; i < multiCharPhonemes.size(); ++i) {
        EXPECT_NE(result[i], multiCharPhonemes[i]);
    }
}

// Test mixed phonemes
TEST(PhonemizeTest, MixedPhonemes) {
    std::vector<std::string> input = {"k", "o", "n", "n", "i", "ch", "i", "w", "a"};
    auto result = piper::mapPhonemes(input);

    EXPECT_EQ(result.size(), input.size());

    // Single char phonemes should remain unchanged
    EXPECT_EQ(result[0], "k");
    EXPECT_EQ(result[1], "o");

    // Multi-char should be mapped
    EXPECT_NE(result[5], "ch");
}

// Test empty input
TEST(PhonemizeTest, EmptyInput) {
    std::vector<std::string> input = {};
    auto result = piper::mapPhonemes(input);

    EXPECT_TRUE(result.empty());
}

// Test single character phonemes
TEST(PhonemizeTest, SingleCharPhonemes) {
    std::vector<std::string> input = {"a", "i", "u", "e", "o"};
    auto result = piper::mapPhonemes(input);

    EXPECT_EQ(result, input); // Should be unchanged
}

// Test Japanese specific phonemes
TEST(PhonemizeTest, JapanesePhonemes) {
    std::vector<std::string> input = {"^", "k", "o", "N", "n", "i", "ch", "i", "w", "a", "$"};
    auto result = piper::mapPhonemes(input);

    // Check markers are preserved
    EXPECT_EQ(result.front(), "^");
    EXPECT_EQ(result.back(), "$");

    // Check "ch" is mapped
    bool found_ch_mapping = false;
    for (size_t i = 0; i < input.size(); ++i) {
        if (input[i] == "ch" && result[i] != "ch") {
            found_ch_mapping = true;
            break;
        }
    }
    EXPECT_TRUE(found_ch_mapping);
}

// Test UTF-8 encoding of PUA characters
TEST(PhonemizeTest, PUAEncodingTest) {
    // Test that PUA characters are properly encoded
    std::vector<std::string> input = {"ch"};
    auto result = piper::mapPhonemes(input);

    ASSERT_EQ(result.size(), 1);

    // Check it's a 3-byte UTF-8 sequence (PUA is in range E000-F8FF)
    const std::string& pua_char = result[0];
    EXPECT_EQ(pua_char.length(), 3); // PUA chars are 3 bytes in UTF-8

    // Check first byte starts with 1110 (0xE0)
    EXPECT_EQ((unsigned char)pua_char[0] & 0xF0, 0xE0);
}

// Main function for test runner
int main(int argc, char **argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}